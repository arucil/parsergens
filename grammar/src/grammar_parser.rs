
pub mod lex;
pub mod ast;

use lalrpop_util::{lalrpop_mod, ParseError};

lalrpop_mod!(pub grammar, "/grammar_parser/grammar.rs");

pub fn parse(
  input: &str
) -> Result<ast::Grammar, ParseError<usize, lex::Token, lex::LexError>> {
  let lexer = lex::Lexer::new(input);
  grammar::DocumentParser::new().parse(input, lexer)
}
