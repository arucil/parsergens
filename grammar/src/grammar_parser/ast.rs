use std::ops::Range;

pub type Grammar = Vec<Spanned<Decl>>;

#[derive(Debug)]
pub enum Decl {
  Token(TokenDecl),
  Start(StartDecl),
  Rule(RuleDecl),
  Skip(SkipDecl),
}

#[derive(Debug)]
pub struct TokenDecl {
  pub name: Spanned<String>,
  pub pattern: TokenPattern,
}

#[derive(Debug)]
pub enum TokenPattern {
  Regex(Spanned<(String, Regex)>),
  String(Spanned<(String, Regex)>),
}

#[derive(Debug)]
pub enum Regex {
  Any,
  Char(char),
  CharSet(Vec<Range<char>>),
  CharClass(CharClassKind),
  Or(Box<Regex>, Box<Regex>),
  Optional(Box<Regex>),
  Many(Box<Regex>),
  Many1(Box<Regex>),
  Seq(Vec<Regex>),
}

#[derive(Debug)]
pub enum CharClassKind {
  Digit,
  Word,
}

#[derive(Debug)]
pub struct SkipDecl {
  pub pattern: TokenPattern,
}

#[derive(Debug)]
pub struct StartDecl {
  pub name: Spanned<String>,
}

#[derive(Debug)]
pub struct RuleDecl {
  pub name: Spanned<String>,
  pub alts: Vec<Spanned<RuleAlt>>,
}

#[derive(Debug)]
pub enum RuleAlt {
  Epsilon,
  Terms(Vec<Spanned<String>>),
}

#[derive(Debug)]
pub struct Spanned<T>(pub Range<usize>, pub T);