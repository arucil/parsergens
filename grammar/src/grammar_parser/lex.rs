use std::collections::VecDeque;
use std::str::CharIndices;
use std::iter::Peekable;
use std::fmt::Display;
use std::fmt;
use super::UserParseError;

pub type Spanned<Tok, Loc, Error> = Result<(Loc, Tok, Loc), Error>;

#[derive(Clone, Copy, Debug)]
pub enum TokenKind {
  Start,
  Token,
  Skip,
  User,

  Ident,
  Regex,
  String,
  CodeBlock,

  Percent,
  Assign,
  Or,
  LParen,
  RParen,
  Comma,
  Colon,
  Asterisk,
  QuestionMark,
  Plus,
  Arrow,

  Enter,
  Leave,
  Separator,
}

#[derive(Clone, Debug)]
pub struct Token<'a> {
  pub kind: TokenKind,
  pub text: &'a str
}

#[derive(Debug)]
pub struct LexError {
  pub kind: LexErrorKind,
  pub span: (usize, usize),
}

#[derive(Debug)]
pub enum LexErrorKind {
  InvalidChar,
  InvalidIndent,
  UnclosedRegex,
  UnclosedString,
  UnclosedCodeBlock,
}

pub struct Lexer<'a> {
  input: &'a str,
  chars: Peekable<CharIndices<'a>>,
  layout_stack: Vec<usize>,
  layout_base: Option<(usize, usize)>,
  buffer: VecDeque<<Self as Iterator>::Item>,
  cur_line_indent: usize,
  column: usize,
  bol: bool,
}

impl<'a> Lexer<'a> {
  pub fn new(input: &'a str) -> Self {
    Self {
      input,
      chars: input.char_indices().peekable(),
      layout_stack: vec![],
      layout_base: None,
      buffer: VecDeque::new(),
      cur_line_indent: 0,
      column: 0,
      bol: true,
    }
  }

  fn advance(&mut self) -> Option<(usize, char)> {
    self.column += 1;
    self.chars.next()
  }

  fn gen_token(&mut self, kind: TokenKind, start: usize, end: usize) {
    if kind.is_layout_start() {
      self.layout_base = Some((self.cur_line_indent, end));
    }

    let token = Token {
      kind,
      text: &self.input[start..end],
    };
    self.buffer.push_back(Ok((start, token, end)));
  }

  fn gen_error(&mut self, kind: LexErrorKind, start: usize, end: usize) {
    let error = LexError {
      kind,
      span: (start, end),
    };
    self.buffer.push_back(Err(UserParseError::LexError(error)));
  }

  fn single_char_token(&mut self, kind: TokenKind, index: usize) {
    self.gen_token(kind, index, index + 1);
  }

  fn lex_quoted(
    &mut self,
    start: usize,
    quote: char,
    kind: TokenKind,
    unclosed_err_kind: LexErrorKind,
  ) {
    let mut unescaped = true;
    let mut end = start;
    loop {
      match self.advance() {
        Some((t, '\\')) => {
          unescaped = !unescaped;
          end += t + 1;
        }
        Some((k, c)) if c == quote && unescaped => {
          break self.gen_token(kind, start, k + 1);
        }
        Some((_, '\n')) | None => {
          break self.gen_error(unclosed_err_kind, start, end);
        }
        Some((t, c)) => {
          unescaped = true;
          end += t + c.len_utf8();
        }
      }
    }
  }

  fn lex_text_block(&mut self, start: usize) {
    let mut lbrace_num = 1;
    while let Some((_, '{')) = self.chars.peek() {
      lbrace_num += 1;
      self.advance();
    }

    let mut rbrace_left = lbrace_num;
    loop {
      if let Some((_, '}')) = self.chars.peek() {
        rbrace_left -= 1;
        self.advance();
      } else if rbrace_left == 0 {
        let end = self.chars.peek().map(|&(i, _)| i).unwrap_or(self.input.len());
        break self.gen_token(TokenKind::CodeBlock, start, end);
      } else if self.chars.next().is_some() {
        rbrace_left = lbrace_num;
      } else {
        break self.gen_error(
          LexErrorKind::UnclosedCodeBlock, start, self.input.len());
      }
    }
  }

  fn gen_tokens(&mut self) -> Option<()> {
    let i = self.chars.peek().map(|&(i, _)| i).unwrap_or(self.input.len());

    // skip whitespace
    while let Some(&(_, c@(' ' | '\n'))) = self.chars.peek() {
      self.advance();
      if c == '\n' {
        self.bol = true;
        self.cur_line_indent = 0;
        self.column = 0;
      } else {
        self.cur_line_indent += 1;
      }
    }

    let (j, c) = if let Some(x) = self.advance() {
      x
    } else {
      // pop all layout on EOF
      while self.layout_stack.pop().is_some() {
        self.gen_token(TokenKind::Leave, i, i);
      }
      if self.buffer.is_empty() {
        return None;
      } else {
        return Some(());
      }
    };

    // skip comment
    if j + 2 <= self.input.len() && &self.input[j..j + 2] == "//" {
      loop {
        let &(_, c) = self.chars.peek()?;
        if c == '\n' {
          return self.gen_tokens();
        } else {
          self.advance();
        }
      }
    }

    if let Some((base_indent, start)) = self.layout_base.take() {
      let end = j;

      if self.column - 1 > base_indent {
        self.layout_stack.push(self.column - 1);

        self.gen_token(TokenKind::Enter, start, end);
      } else {
        self.gen_error(LexErrorKind::InvalidIndent, start, end);
      }
    } else if self.bol {
      while let Some(&indent) = self.layout_stack.last() {
        if self.cur_line_indent >= indent {
          break;
        }
        self.layout_stack.pop();
        self.gen_token(TokenKind::Leave, i, i);
      }

      let base_indent = self.layout_stack.last().cloned().unwrap_or(0);
      if self.cur_line_indent == base_indent {
        let start = i;
        let end = j;
        self.gen_token(TokenKind::Separator, start, end);
      } else {
        // do nothing
      }
    }

    self.bol = false;

    match c {
      '/' => {
        self.lex_quoted(j, '/', TokenKind::Regex, LexErrorKind::UnclosedRegex);
      }
      '"' => {
        self.lex_quoted(j, '"', TokenKind::String, LexErrorKind::UnclosedString);
      }
      '%' => self.single_char_token(TokenKind::Percent, j),
      '=' => self.single_char_token(TokenKind::Assign, j),
      '|' => self.single_char_token(TokenKind::Or, j),
      ',' => self.single_char_token(TokenKind::Comma, j),
      ':' => self.single_char_token(TokenKind::Colon, j),
      '*' => self.single_char_token(TokenKind::Asterisk, j),
      '+' => self.single_char_token(TokenKind::Plus, j),
      '?' => self.single_char_token(TokenKind::QuestionMark, j),
      '(' => self.single_char_token(TokenKind::LParen, j),
      ')' => self.single_char_token(TokenKind::RParen, j),
      '-' => {
        if let Some((_, '>')) = self.chars.peek() {
          self.advance();
          self.gen_token(TokenKind::Arrow, j, j + 2);
        } else {
          self.gen_error(LexErrorKind::InvalidChar, j, j + 1);
        }
      }
      '{' => {
        self.lex_text_block(j);
      }
      _ if c.is_ascii_alphabetic() || c == '\'' => {
        let mut last = j;
        loop {
          match self.chars.peek() {
            Some(&(i, c)) => {
              if !(c.is_alphanumeric() || c == '-' || c == '\'' || c == '_') {
                break;
              }
              last = i;
            }
            None => break,
          }
          self.advance();
        }

        let text = &self.input[j..last + 1];
        let kind = match text {
          "start" => TokenKind::Start,
          "token" => TokenKind::Token,
          "skip" => TokenKind::Skip,
          "user" => TokenKind::User,
          _ => TokenKind::Ident,
        };

        self.gen_token(kind, j, last + 1);
      }
      _ => {
        self.gen_error(LexErrorKind::InvalidChar, j, j + c.len_utf8());
      }
    }

    Some(())
  }
}

impl<'a> Iterator for Lexer<'a> {
  type Item = Spanned<Token<'a>, usize, UserParseError>;

  fn next(&mut self) -> Option<Self::Item> {
    while self.buffer.is_empty() {
      self.gen_tokens()?;
    }
    self.buffer.pop_front()
  }
}

impl<'a> Display for Token<'a> {
  fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
    write!(f, "<{:?}: {:?}>", self.kind, self.text)
  }
}

impl TokenKind {
  fn is_layout_start(&self) -> bool {
    match self {
      Self::Arrow => true,
      _ => false,
    }
  }
}

#[cfg(test)]
mod tests {
  use super::*;
  use insta::assert_debug_snapshot;

  #[test]
  fn tokens() {
    let result = Lexer::new(r#"%% web /\d+\\[0-9]\//  //12345
    | token,start  

  =  ( Ax3_ ) "abcdef" foo-bar''
   
    "#).collect::<Vec<_>>();

    assert_debug_snapshot!(result);
  }

  #[test]
  fn code_block() {
    let result = Lexer::new(r#"
{ // not comment }  {{ } not end }}
  {{{ }} }} } }}}
   {{}}}} }}"#).collect::<Vec<_>>();

    assert_debug_snapshot!(result);
  }

  #[test]
  fn layout_nested() {
    let result = Lexer::new(r#"
foo ->
  "bar"-> ( = )
          
          %
    +
  * //
    "#).collect::<Vec<_>>();

    assert_debug_snapshot!(result);
  }

  #[test]
  fn layout_multiple() {
    let result = Lexer::new(r#"
-> -> -> foo
         bar ()
      baz
   qux
lorem
  ipsum
 dolor

->
    sit
    amet
    "#).collect::<Vec<_>>();

    assert_debug_snapshot!(result);
  }

  #[test]
  fn separator_toplevel() {
    let result = Lexer::new(r#"
foo( )
bar "a"
baz
    "#).collect::<Vec<_>>();

    assert_debug_snapshot!(result);
  }
}