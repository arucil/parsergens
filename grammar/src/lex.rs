use std::str::CharIndices;
use std::iter::Peekable;

pub type Spanned<Tok, Loc, Error> = Result<(Loc, Tok, Loc), Error>;

#[derive(Clone, Debug)]
pub enum TokenKind {
  Ident,
  Start,
  Token,
  Regex,

  Percent,
  Assign,
  Or,
  LParen,
  RParen,
  Comma,

  Indent,
  Newline,
}

#[derive(Clone, Debug)]
pub struct Token<'a> {
  pub kind: TokenKind,
  pub text: &'a str
}

#[derive(Debug)]
pub enum LexError {
  UnclosedRegex,
  InvalidChar,
}

pub struct Lexer<'a> {
  input: &'a str,
  chars: Peekable<CharIndices<'a>>,
  bol: bool,
}

impl<'a> Lexer<'a> {
  pub fn new(input: &'a str) -> Self {
    Self {
      input,
      chars: input.char_indices().peekable(),
      bol: true
    }
  }

  fn single_char_token(
    &self, kind: TokenKind, index: usize
  ) -> Option<<Self as Iterator>::Item> {
    let token = Token {
      kind,
      text: &self.input[index..index + 1],
    };
    return Some(Ok((index, token, index + 1)));
  }
}

impl<'a> Iterator for Lexer<'a> {
  type Item = Spanned<Token<'a>, usize, LexError>;

  fn next(&mut self) -> Option<Self::Item> {
    let &(i, _) = self.chars.peek()?;

    while let Some((_, ' ')) = self.chars.peek() {
      self.chars.next();
    }

    let &(j, c) = self.chars.peek()?;
    if j > i && self.bol {
      if c != '\n' && &self.input[j..j + 2] != "//" {
        let tok = Token {
          kind: TokenKind::Indent,
          text: &self.input[i..j],
        };
        return Some(Ok((i, tok, j)));
      }
    }

    self.bol = false;
    self.chars.next();

    match c {
      '\n' => {
        // collapse newlines
        if j > i || self.bol {
          return self.next();
        }

        self.bol = true;
        let token = Token {
          kind: TokenKind::Newline,
          text: &self.input[j..j + 1],
        };
        return Some(Ok((j, token, j + 1)));
      }
      '/' => {
        if let Some((_, '/')) = self.chars.peek() {
          self.chars.next();
          loop {
            let (k, c) = self.chars.next()?;
            if c == '\n' {
              self.bol = true;
              let token = Token {
                kind: TokenKind::Newline,
                text: &self.input[k..k + 1],
              };
              return Some(Ok((k, token, k + 1)));
            }
          }
        } else {
          let mut unescaped = true;
          loop {
            match self.chars.next() {
              Some((_, '\\')) => {
                unescaped = !unescaped;
              }
              Some((k, '/')) if unescaped => {
                self.chars.next();
                let token = Token {
                  kind: TokenKind::Regex,
                  text: &self.input[j..k + 1],
                };
                return Some(Ok((j, token, k + 1)));
              }
              None => {
                return Some(Err(LexError::UnclosedRegex))
              }
              _ => {
                unescaped = true;
              }
            }
          }
        }
      }
      '%' => return self.single_char_token(TokenKind::Percent, j),
      '=' => return self.single_char_token(TokenKind::Assign, j),
      '|' => return self.single_char_token(TokenKind::Or, j),
      ',' => return self.single_char_token(TokenKind::Comma, j),
      '(' => return self.single_char_token(TokenKind::LParen, j),
      ')' => return self.single_char_token(TokenKind::RParen, j),
      _ if c.is_ascii_alphabetic() => {
        let mut last = j;
        loop {
          match self.chars.peek() {
            Some(&(i, c)) => {
              if !(c.is_alphanumeric() || c == '_') {
                break;
              }
              last = i;
            }
            None => break,
          }
          self.chars.next();
        }

        let text = &self.input[j..last + 1];
        let kind = match text {
          "start" => TokenKind::Start,
          "token" => TokenKind::Token,
          _ => TokenKind::Ident,
        };

        let token = Token {
          kind,
          text,
        };
        return Some(Ok((j, token, last + 1)));
      }
      _ => {}
    }

    Some(Err(LexError::InvalidChar))
  }
}

#[cfg(test)]
mod tests {
  use super::*;
  use insta::assert_debug_snapshot;

  #[test]
  fn tokens() {
    let result = Lexer::new(r"
%% web /\d+[0-9]\//  //12345
    | token,start  

  =  ( Ax3_ )
   
    ").collect::<Vec<_>>();

    assert_debug_snapshot!(result);
  }
}