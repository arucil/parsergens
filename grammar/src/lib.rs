#![feature(or_patterns, never_type)]

use lalrpop_util::ParseError;
use codespan_reporting::diagnostic::{Diagnostic, Label};
use codespan_reporting::term;
use codespan_reporting::term::termcolor::{ColorChoice, StandardStream};
use codespan_reporting::files::SimpleFile;

mod grammar_parser;

pub mod lexer;
pub mod grammar;

pub use crate::grammar::*;
pub use lexer::{Lexer, TokenId, TokenIdGen};

use grammar_parser::ast;
use grammar_parser::lex::{Token, LexErrorKind, LexError};
use grammar_parser::regex::{RegexError, RegexErrorKind};
use grammar_parser::UserParseError;
use lexer::LexerError;

#[cfg(not(debug_assertions))]
use std::collections::{HashMap, HashSet};
#[cfg(not(debug_assertions))]
use bimap::BiHashMap;

#[cfg(debug_assertions)]
use indexmap::{IndexMap, IndexSet};
#[cfg(debug_assertions)]
use bimap::BiBTreeMap;

#[cfg(not(debug_assertions))]
type Map<K, V> = HashMap<K, V>;

#[cfg(debug_assertions)]
type Map<K, V> = IndexMap<K, V>;

#[cfg(not(debug_assertions))]
type BiMap<K, V> = BiHashMap<K, V>;

#[cfg(debug_assertions)]
type BiMap<K, V> = BiBTreeMap<K, V>;

#[cfg(not(debug_assertions))]
type Set<K> = HashSet<K>;

#[cfg(debug_assertions)]
type Set<K> = IndexSet<K>;


#[derive(Debug)]
pub struct GrammarError {
  kind: GrammarErrorKind,
  message: String,
  span: (usize, usize),
}

#[derive(Debug)]
pub enum GrammarErrorKind {
  ParseError,
  SymbolNotFound,
  NameConflict,
}

pub fn build(grammar: &str) -> Result<Grammar, GrammarError> {
  let ast = grammar_parser::parse(grammar)
    .map_err(|err| err.into())?;

  let decls = ast.iter()
    .filter_map(|decl| match &decl.1 {
      ast::Decl::Token(decl) => Some(decl),
      _ => None,
    })
    .collect::<Vec<_>>();

  let skips = ast.iter()
    .filter_map(|decl| match &decl.1 {
      ast::Decl::Skip(decl) => Some(decl),
      _ => None,
    })
    .collect::<Vec<_>>();

  let starts = ast.iter()
    .filter_map(|decl| match &decl.1 {
      ast::Decl::Start(decl) => Some(decl),
      _ => None,
    })
    .collect::<Vec<_>>();

  let rules = ast.iter()
    .filter_map(|decl| match &decl.1 {
      ast::Decl::Rule(decl) => Some(decl),
      _ => None,
    })
    .collect::<Vec<_>>();

  let lexer = Lexer::new(&decls, &skips).map_err(|err| {
    match err {
      LexerError::NoTokens => {
        GrammarError {
          kind: GrammarErrorKind::ParseError,
          message: format!("no token declarations"),
          span: (0, grammar.len()),
        }
      }
      LexerError::RegexError(err) => {
        err.into()
      }
    }
  })?;

  let mut nt_id_gen = NonterminalIdGen::default();
  let nts = rules.iter()
    .map(|decl| {
      if lexer.tokens.get_by_right(&decl.name.1).is_some() {
        return Err(GrammarError {
          kind: GrammarErrorKind::NameConflict,
          message: format!("rule name collides with token name"),
          span: decl.name.0,
        });
      }

      let id = nt_id_gen.gen();

      Ok((id, decl.name.1.clone()))
    })
    .collect::<Result<BiMap<_, _>, GrammarError>>()?;

  let start_nts = starts.iter()
    .map(|decl| {
      if let Some(&id) = nts.get_by_right(&decl.name.1) {
        Ok(id)
      } else {
        Err(GrammarError {
          kind: GrammarErrorKind::SymbolNotFound,
          message: format!("nonterminal not found"),
          span: decl.name.0,
        })
      }
    })
    .collect::<Result<Set<_>, GrammarError>>()?;

  let mut prods = vec![];
  let mut nt_prods = Map::new();

  for rule in &rules {
    let nt = *nts.get_by_right(&rule.name.1).unwrap();
    let start = prods.len();

    for alt in &rule.alts {
      let items = match &alt.1 {
        ast::RuleAlt::Epsilon => vec![],
        ast::RuleAlt::Terms(terms) => {
          convert_terms(terms, &nts, &lexer.tokens)?
        }
      };

      prods.push(Production {
        nt,
        items,
      });
    }

    nt_prods.insert(nt, start..prods.len());
  }

  Ok(Grammar {
    prods,
    start_nts,
    nts,
    nt_prods,
    lexer,
  })
}

fn convert_terms(
  terms: &[ast::Term],
  nts: &BiMap<NonterminalId, String>,
  tokens: &BiMap<TokenId, String>,
) -> Result<Vec<Item>, GrammarError> {
  let mut items = vec![];

  for term in terms {
    match term {
      ast::Term::Symbol(sym) => {
        if let Some(&id) = nts.get_by_right(&sym.1) {
          items.push(Item::Nonterminal(id));
        } else if let Some(&id) = tokens.get_by_right(&sym.1) {
          items.push(Item::Token(id));
        } else {
          return Err(GrammarError {
            kind: GrammarErrorKind::SymbolNotFound,
            message: format!("nonterminal or token not found"),
            span: sym.0,
          });
        }
      }
      ast::Term::Optional(terms) => {
        items.push(Item::Optional(convert_terms(terms, nts, tokens)?));
      }
      ast::Term::Many(terms) => {
        items.push(Item::Many(convert_terms(terms, nts, tokens)?));
      }
      ast::Term::Many1(terms) => {
        items.push(Item::Many1(convert_terms(terms, nts, tokens)?));
      }
    }
  }

  Ok(items)
}

pub fn report_error(input: &str, error: &GrammarError) {
  let writer = StandardStream::stderr(ColorChoice::Always);
  let config = term::Config::default();
  let files = SimpleFile::new("", input);

  let diagnostic = Diagnostic::error();
  let diagnostic = match error.kind {
    GrammarErrorKind::ParseError => {
      diagnostic.with_message("syntax error")
    }
    GrammarErrorKind::NameConflict => {
      diagnostic.with_message("name conflict")
    }
    GrammarErrorKind::SymbolNotFound => {
      diagnostic.with_message("symbol not found")
    }
  };
  let diagnostic = diagnostic.with_labels(vec![
    Label::primary((), error.span.0..error.span.1).with_message(&error.message)
  ]);

  term::emit(&mut writer.lock(), &config, &files, &diagnostic).unwrap();
}

impl<'a> Into<GrammarError> for ParseError<usize, Token<'a>, UserParseError> {
  fn into(self) -> GrammarError {
    match self {
      Self::InvalidToken { location } => {
        GrammarError {
          kind: GrammarErrorKind::ParseError,
          message: format!("invalid token or EOF"),
          span: (location, location),
        }
      }
      Self::UnrecognizedEOF { location, expected } => {
        GrammarError {
          kind: GrammarErrorKind::ParseError,
          message: format!("expected {}, found EOF", expected.join(", ")),
          span: (location, location),
        }
      }
      Self::UnrecognizedToken { token, expected } => {
        GrammarError {
          kind: GrammarErrorKind::ParseError,
          message: format!("expected {}, found {}", expected.join(", "), token.1),
          span: (token.0, token.2),
        }
      }
      Self::ExtraToken { token } => {
        GrammarError {
          kind: GrammarErrorKind::ParseError,
          message: token.1.to_string(),
          span: (token.0, token.2),
        }
      }
      Self::User { error } => {
        match error {
          UserParseError::LexError(error) => error.into(),
          UserParseError::RegexError(error) => error.into(),
        }
      }
    }
  }
}

impl<'a> Into<GrammarError> for LexError {
  fn into(self) -> GrammarError {
    match self.kind {
      LexErrorKind::UnclosedRegex => {
        GrammarError {
          kind: GrammarErrorKind::ParseError,
          message: format!("unclosed regex"),
          span: self.span,
        }
      }
      LexErrorKind::UnclosedString => {
        GrammarError {
          kind: GrammarErrorKind::ParseError,
          message: format!("unclosed string"),
          span: self.span,
        }
      }
      LexErrorKind::InvalidChar => {
        GrammarError {
          kind: GrammarErrorKind::ParseError,
          message: format!("invalid character"),
          span: self.span,
        }
      }
    }
  }
}

impl<'a> Into<GrammarError> for RegexError {
  fn into(self) -> GrammarError {
    match self.kind {
      RegexErrorKind::SyntaxError => {
        GrammarError {
          kind: GrammarErrorKind::ParseError,
          message: format!("regex syntax error"),
          span: self.span,
        }
      }
      RegexErrorKind::InvalidCodePoint => {
        GrammarError {
          kind: GrammarErrorKind::ParseError,
          message: format!("invalid code point"),
          span: self.span,
        }
      }
      RegexErrorKind::Empty => {
        GrammarError {
          kind: GrammarErrorKind::ParseError,
          message: format!("regex or string may accept empty string"),
          span: self.span,
        }
      }
    }
  }
}

#[cfg(test)]
mod tests {
  use super::*;
  use insta::assert_debug_snapshot;

  #[test]
  fn expr() {
    let input = r#"
%start expr

%token PLUS "+"
%token MINUS "-"
%token MUL "*"
%token DIV "/"
%token LPAREN "("
%token RPAREN ")"
%token COMMA ","
%token NUMBER /\d+(\.\d*)?/
%token IDENT /[a-zA-Z][\w_]*/

%skip /[ \n]/

expr = expr PLUS expr
  | expr MINUS expr
  | expr MUL expr
  | expr DIV expr
  | factor

factor =
    NUMBER
  | LPAREN expr RPAREN
  | MINUS factor
  | IDENT
  | IDENT LPAREN param_list RPAREN

param_list =
    ()
  | expr ( COMMA expr )*
    "#;

    assert_debug_snapshot!(build(input));
  }
}