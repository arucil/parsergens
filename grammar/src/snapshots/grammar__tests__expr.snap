---
source: grammar/src/lib.rs
expression: build(input)
---
Ok(
    Grammar {
        rules: {
            RuleId(
                0,
            ): Rule {
                name: "expr",
                alts: [
                    [
                        Rule(
                            RuleId(
                                0,
                            ),
                        ),
                        Token(
                            TokenId(
                                0,
                            ),
                        ),
                        Rule(
                            RuleId(
                                0,
                            ),
                        ),
                    ],
                    [
                        Rule(
                            RuleId(
                                0,
                            ),
                        ),
                        Token(
                            TokenId(
                                1,
                            ),
                        ),
                        Rule(
                            RuleId(
                                0,
                            ),
                        ),
                    ],
                    [
                        Rule(
                            RuleId(
                                0,
                            ),
                        ),
                        Token(
                            TokenId(
                                2,
                            ),
                        ),
                        Rule(
                            RuleId(
                                0,
                            ),
                        ),
                    ],
                    [
                        Rule(
                            RuleId(
                                0,
                            ),
                        ),
                        Token(
                            TokenId(
                                3,
                            ),
                        ),
                        Rule(
                            RuleId(
                                0,
                            ),
                        ),
                    ],
                    [
                        Rule(
                            RuleId(
                                1,
                            ),
                        ),
                    ],
                ],
            },
            RuleId(
                1,
            ): Rule {
                name: "factor",
                alts: [
                    [
                        Token(
                            TokenId(
                                7,
                            ),
                        ),
                    ],
                    [
                        Token(
                            TokenId(
                                4,
                            ),
                        ),
                        Rule(
                            RuleId(
                                0,
                            ),
                        ),
                        Token(
                            TokenId(
                                5,
                            ),
                        ),
                    ],
                    [
                        Token(
                            TokenId(
                                1,
                            ),
                        ),
                        Token(
                            TokenId(
                                7,
                            ),
                        ),
                    ],
                    [
                        Token(
                            TokenId(
                                8,
                            ),
                        ),
                    ],
                    [
                        Token(
                            TokenId(
                                8,
                            ),
                        ),
                        Token(
                            TokenId(
                                4,
                            ),
                        ),
                        Rule(
                            RuleId(
                                2,
                            ),
                        ),
                        Token(
                            TokenId(
                                5,
                            ),
                        ),
                    ],
                    [
                        Token(
                            TokenId(
                                8,
                            ),
                        ),
                        Token(
                            TokenId(
                                4,
                            ),
                        ),
                        Token(
                            TokenId(
                                5,
                            ),
                        ),
                    ],
                ],
            },
            RuleId(
                2,
            ): Rule {
                name: "param_list",
                alts: [
                    [
                        Rule(
                            RuleId(
                                0,
                            ),
                        ),
                    ],
                    [
                        Rule(
                            RuleId(
                                2,
                            ),
                        ),
                        Token(
                            TokenId(
                                6,
                            ),
                        ),
                        Rule(
                            RuleId(
                                0,
                            ),
                        ),
                    ],
                ],
            },
        },
        start_rules: {
            RuleId(
                0,
            ),
        },
        rule_names: {
            RuleId(
                0,
            ) <> "expr",
            RuleId(
                2,
            ) <> "param_list",
            RuleId(
                1,
            ) <> "factor",
        },
        lexer: Lexer {
            dfa: Dfa {
                start: 1,
                transitions: {
                    (
                        State(1),
                        7,
                    ): State(2),
                    (
                        State(1),
                        9,
                    ): State(3),
                    (
                        State(1),
                        6,
                    ): State(4),
                    (
                        State(1),
                        11,
                    ): State(5),
                    (
                        State(1),
                        4,
                    ): State(6),
                    (
                        State(1),
                        5,
                    ): State(7),
                    (
                        State(1),
                        8,
                    ): State(8),
                    (
                        State(1),
                        18,
                    ): State(9),
                    (
                        State(1),
                        14,
                    ): State(9),
                    (
                        State(1),
                        2,
                    ): State(10),
                    (
                        State(1),
                        0,
                    ): State(10),
                    (
                        State(1),
                        12,
                    ): State(11),
                    (
                        State(9),
                        12,
                    ): State(12),
                    (
                        State(9),
                        18,
                    ): State(12),
                    (
                        State(9),
                        14,
                    ): State(12),
                    (
                        State(9),
                        16,
                    ): State(12),
                    (
                        State(11),
                        12,
                    ): State(13),
                    (
                        State(11),
                        10,
                    ): State(14),
                    (
                        State(12),
                        12,
                    ): State(12),
                    (
                        State(12),
                        18,
                    ): State(12),
                    (
                        State(12),
                        14,
                    ): State(12),
                    (
                        State(12),
                        16,
                    ): State(12),
                    (
                        State(13),
                        12,
                    ): State(13),
                    (
                        State(13),
                        10,
                    ): State(14),
                    (
                        State(14),
                        12,
                    ): State(15),
                    (
                        State(15),
                        12,
                    ): State(15),
                },
                accept_states: {
                    State(2): TokenId(
                        0,
                    ),
                    State(3): TokenId(
                        1,
                    ),
                    State(4): TokenId(
                        2,
                    ),
                    State(5): TokenId(
                        3,
                    ),
                    State(6): TokenId(
                        4,
                    ),
                    State(7): TokenId(
                        5,
                    ),
                    State(8): TokenId(
                        6,
                    ),
                    State(9): TokenId(
                        8,
                    ),
                    State(10): TokenId(
                        9,
                    ),
                    State(11): TokenId(
                        7,
                    ),
                    State(12): TokenId(
                        8,
                    ),
                    State(13): TokenId(
                        7,
                    ),
                    State(14): TokenId(
                        7,
                    ),
                    State(15): TokenId(
                        7,
                    ),
                },
            },
            char_intervals: [
                10,
                11,
                32,
                33,
                40,
                41,
                42,
                43,
                44,
                45,
                46,
                47,
                48,
                58,
                65,
                91,
                95,
                96,
                97,
                123,
            ],
            token_names: {
                TokenId(
                    0,
                ) <> "PLUS",
                TokenId(
                    2,
                ) <> "MUL",
                TokenId(
                    8,
                ) <> "IDENT",
                TokenId(
                    3,
                ) <> "DIV",
                TokenId(
                    7,
                ) <> "NUMBER",
                TokenId(
                    6,
                ) <> "COMMA",
                TokenId(
                    4,
                ) <> "LPAREN",
                TokenId(
                    1,
                ) <> "MINUS",
                TokenId(
                    5,
                ) <> "RPAREN",
            },
        },
    },
)
